### PR TITLE
PKI - Honor header If-Modified-Since if present

### PR DESCRIPTION
**Pull request by Gabrielopesantos**

### CODE REVIEW

**Comment by hsimon-hashicorp** [comment]
Hi @Gabrielopesantos - please don't forget to add a changelog entry too. :) Please see the [contributing.md](https://github.com/hashicorp/vault/blob/main/CONTRIBUTING.md) for more details.

**Comment by cipherboy** [comment]
Sorry @Gabrielopesantos -- I've been on vacation for a bit, I'll take a look at this week now that I'm back :)

File: builtin/logical/pki/path_fetch.go [hunk 1]
```diff
@@ -179,6 +181,19 @@ func (b *backend) pathFetchRead(ctx context.Context, req *logical.Request, data
 			contentType = "application/pkix-cert"
 		}
 	case req.Path == "crl" || req.Path == "crl/pem":
+		if hasHeader(headerIfModifiedSince, req) {
+            // Maybe a sc declared at the top of function?
```
Comment (cipherboy) [inline builtin/logical/pki/path_fetch.go:185]: Yeah I like hoisting this up and moving it earlier since the other branch also uses this, yeah.

File: builtin/logical/pki/path_fetch.go [hunk 2]
```diff
@@ -324,6 +339,19 @@ reply:
 		return
 	case response.IsError():
 		return response, nil
+	case httpStatusCode != 0:
+		response = &logical.Response{
+			Data: map[string]interface{}{
+				logical.HTTPContentType: contentType,
+				logical.HTTPRawBody:     certificate,
```
Comment (cipherboy) [inline builtin/logical/pki/path_fetch.go:346]: I think we want to unset this if the status code is 304? Otherwise I believe clients will see the whole response, which I don't think is correct? We should probably test and validate this against say, curl or something. 

I think https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/304#compatibility_notes and https://www.webfx.com/web-development/glossary/http-status-codes/what-is-a-304-status-code/ backs up not sending a body in this case.

**Review by cipherboy — COMMENTED** [review]
This looks great! I think you'd want to do it for `pathGetIssuerCRL` and perhaps `pathGetRawIssuer` as well? :-) You might need to store the creation time on the `IssuerEntry` object as well, perhaps migrating if one doesn't exist. I wouldn't do the CA if-modified-since on the default issuer (`path_fetch.go`).

File: builtin/logical/pki/path_fetch.go (hunk 2 above)
Comment (Gabrielopesantos) [reply builtin/logical/pki/path_fetch.go:346]: I used curl during the development to test the endpoint, from what I understand, with the current implementation and since these values are still default when set, the body is empty and content-type isn't returned either.
I have now tried to unset both values, and if content-type isn't set, the response is 500. (Body seems to be optional).

Comment (cipherboy) [reply builtin/logical/pki/path_fetch.go:346]: Ah interesting. That's fine then :-) I hadn't realized the body wouldn't be sent if we had specified it in the response object. Kinda weird tbh.

File: builtin/logical/pki/path_fetch.go (hunk 1 above)
Comment (Gabrielopesantos) [reply builtin/logical/pki/path_fetch.go]: > This looks great! I think you'd want to do it for `pathGetIssuerCRL` and perhaps `pathGetRawIssuer` as well? :-) You might need to store the creation time on the `IssuerEntry` object as well, perhaps migrating if one doesn't exist. I wouldn't do the CA if-modified-since on the default issuer (`path_fetch.go`).

Hey, looked at both of the functions/endpoints? that you mentioned but I don't really understand why we would want to also add the IfModifiedSince check in `pathGetRawIssuer` as it does not seem to cover anything related to CRL.

Comment (cipherboy) [reply builtin/logical/pki/path_fetch.go]: Ah, I can explain that better, but I'll leave it up to you if you want to do it. I'm happy with the scope of this PR if you don't. :-) 

So, the original request (#3190 linked above) said they had two use cases:

1. Auto-deploying the CRL.
2. Auto-deploying the CA. 

They want to avoid the "fetch and compare" with the existing approach (without these headers) and have Vault indicate if something has changed.

While you have addressed the first one (thanks!) the second one is a bit more subtle: we need to know not only the generation date of the CA (which isn't the same as the NotBefore/NotAfter on the cert) -- but also the last time the default was modified.

I was thinking that (in my original comment), as a substitute/proxy for the later (fetching `/cert/ca`), we could add this Header logic to `/issuer/:issuer_ref/{json,der,pem}` -- there, we'd need some field as a modification date. My suggestion was to add the generation date on the IssuerEntry to compare against...

However, this doesn't really make sense -- when fetching by issuer ID, the value of the CA there will never change. But when fetching by name, the generation date isn't the correct field, but I think instead a modification timestamp is the correct field.

If we had that field, we could update the IssuerEntry's LastModifiedTimestamp on both name change and on default ref change, and that'd allow the global `/cert/ca` or `/issuer/default` to also be fetched with the header.

Then, you can use the IfModifiedSince header to detect if the issuer has been rotated somehow, and only rarely have the same cert fetched (if you changed its name but its still default for instance). 

Let me know if that makes more sense? :-)

Comment (Gabrielopesantos) [reply builtin/logical/pki/path_fetch.go]: Hello, I am aware of the original request but I thought, given your initial message, it was not necessary to do it for the CA endpoint(s). I don't mind including it as part of this PR just thought it was not the focus.
Regarding your explanation, it made sense and helped me create an initial version of the "check if modified" for the CA endpoints.
I still intend to take another look to see if there is a better way to do it, but feel free to take a look at it and drop some feedback. Thanks

Comment (cipherboy) [reply builtin/logical/pki/path_fetch.go]: Sorry Gabriel! I was out on PTO again when you replied. I'll take a look today. :-)

File: builtin/logical/pki/config_util.go [hunk 3]
```diff
@@ -44,9 +45,17 @@ func (sc *storageContext) updateDefaultIssuerId(id issuerID) error {
...
-		return sc.setIssuersConfig(&issuerConfigEntry{
+		err := sc.setIssuersConfig(&issuerConfigEntry{
 			DefaultIssuerId: id,
 		})
+
+		if err != nil {
+			return err
+		}
+
+		// Can we assume that this fetch won't raise an error?
```
Comment (cipherboy) [inline builtin/logical/pki/config_util.go:56]: Hmmm... I'd prefer not to. Easy enough to return the err if it does occur. Otherwise, we'd get a nil issuer pointer to try to set the last modified timestamp at... The entry was already written so the main change will persist either way.

File: builtin/logical/pki/path_fetch_issuers.go [hunk 4]
```diff
@@ -92,7 +93,7 @@ issuer's computed CAChain field, when non-empty.`,
 		Description: `Behavior of leaf's NotAfter fields: "err" to error
 if the computed NotAfter date exceeds that of this issuer; "truncate" to
 silently truncate to that of this issuer; or "permit" to allow this
-issuance to succeed (with NotAfter exceeding that of an issuer). Note that
+        issuance to succeed (with NotAfter exceeding that of an issuer). Note that
```
Comment (cipherboy) [inline builtin/logical/pki/path_fetch_issuers.go:96]: Just seems like an accidental change

```suggestion
issuance to succeed (with NotAfter exceeding that of an issuer). Note that
```

File: builtin/logical/pki/path_fetch_issuers.go [hunk 5]
```diff
@@ -262,8 +263,10 @@ func (b *backend) pathUpdateIssuer(ctx context.Context, req *logical.Request, da
 
 	var oldName string
 	if newName != issuer.Name {
+		// If name is updated, we update issuerEntry LastModified?
```
Comment (cipherboy) [inline builtin/logical/pki/path_fetch_issuers.go:266]: I agree this is correct. If we've updated two issuers two cross their names (old-root <-> new root or some such) -- you'd want to inform clients.

Better to make clients fetch the new cert a few extra times than hide changes from them.

File: builtin/logical/pki/path_fetch_issuers.go [hunk 6]
```diff
@@ -509,9 +514,31 @@ func (b *backend) pathGetRawIssuer(ctx context.Context, req *logical.Request, da
...
+
+	// tmp
+	responseHeaders := map[string][]string{}
+	if hasHeader(headerIfModifiedSince, req) {
+		before, err := sc.isIfModifiedSinceBeforeIssuerLastModified, ref, responseHeaders)
+		if err != nil {
+			return nil, err // errorResponse, nil or nil, err?
+		}
+		if before {
+			return &logical.Response{
```
Comment (cipherboy) [inline builtin/logical/pki/path_fetch_issuers.go:528]: Wondering if we want to refactor to make a "not modified since" generic response minter? Just a thought :-)

File: builtin/logical/pki/path_fetch_issuers.go [hunk 7]
```diff
@@ -509,9 +514,31 @@ func (b *backend) pathGetRawIssuer(ctx context.Context, req *logical.Request, da
...
 	}
 
-	certificate := []byte(issuer.Certificate)
-
 	var contentType string
+	var certificate []byte
+
+	// tmp
+	responseHeaders := map[string][]string{}
+	if hasHeader(headerIfModifiedSince, req) {
```
Comment (cipherboy) [inline builtin/logical/pki/path_fetch_issuers.go:522]: If we do do that refactor, I wonder if we could put all of this in a helper? `resp, return := sendNotModifiedResponseIfNecessary(...)`

File: builtin/logical/pki/path_fetch_issuers.go [hunk 8]
```diff
@@ -509,9 +514,31 @@ func (b *backend) pathGetRawIssuer(ctx context.Context, req *logical.Request, da
...
+		before, err := sc.isIfModifiedSinceBeforeIssuerLastModified, ref, responseHeaders)
+		if err != nil {
+			return nil, err // errorResponse, nil or nil, err?
```
Comment (cipherboy) [inline builtin/logical/pki/path_fetch_issuers.go:525]: Its an internal error so I'm fine with `return nil, err`. 

I _think_ `errorResponse` is mostly for bad user inputs. @stevendpclark ?

File: builtin/logical/pki/path_fetch_issuers.go [hunk 9]
```diff
@@ -679,7 +706,29 @@ func (b *backend) pathGetIssuerCRL(ctx context.Context, req *logical.Request, da
...
+				Data: map[string]interface{}{
+					logical.HTTPContentType: contentType,
+					logical.HTTPRawBody:     certificate,
+					logical.HTTPStatusCode:  304,
+				},
+				Headers: responseHeaders,
+			}, nil
+		}
+	}
+	// tmp
```
Comment (cipherboy) [inline builtin/logical/pki/path_fetch_issuers.go:731]: You know this, but I'm just making sure I don't accidentally merge this with `// tmp` comments left in :-)

**Review by cipherboy — COMMENTED** [review]
Many thanks! I think this looks great! :-) 

A few comments inline, perhaps @stevendpclark could take a look too. But I think this is ready to unmark from draft and we'll merge it.

I've think we've got some changes in the pipeline that'll probably cause merge conflicts -- would you like us to resolve those or would you like to? If we merge this sooner rather than later, we can definitely make this release and perhaps avoid some of those conflicts... :-D 


Happy to write the tests if you don't want to, too. :-)

File: builtin/logical/pki/path_fetch_issuers.go (hunk 8 above)
Comment (stevendpclark) [reply builtin/logical/pki/path_fetch_issuers.go:525]: Yes that is my understanding.

File: builtin/logical/pki/storage.go [hunk 10]
```diff
@@ -511,6 +518,9 @@ func (sc *storageContext) fetchIssuerById(issuerId issuerID) (*issuerEntry, erro
 
 func (sc *storageContext) writeIssuer(issuer *issuerEntry) error {
 	issuerId := issuer.ID
+	if issuer.LastModified.IsZero() {
+		issuer.LastModified = time.Now().In(time.FixedZone("GMT", 0))
```
[the remark this answers is not in the document]
Comment (Gabrielopesantos) [reply builtin/logical/pki/storage.go:522]: From what I tested and understood, if we let the value being set unconditionally, as `writeIssuer` is called for every issuer in `rebuildIssuersChains`, we are going to update values that were previously set.

File: builtin/logical/pki/config_util.go [hunk 11]
```diff
@@ -44,9 +45,17 @@ func (sc *storageContext) updateDefaultIssuerId(id issuerID) error {
...
+		issuer, _ := sc.fetchIssuerById(id)
+		issuer.LastModified = time.Now().In(time.FixedZone("GMT", 0))
```
[the remark this answers is not in the document]
Comment (Gabrielopesantos) [reply builtin/logical/pki/config_util.go:58]: I am not too sure but I don't think so.

[the remark this answers is not in the document]
Comment (Gabrielopesantos) [reply builtin/logical/pki/config_util.go:60]: I don't think so, I have run all my manual tests without it and everything worked. Is what's commented below what you mean with the question?

**Comment by Gabrielopesantos** [comment]
Ok, with this last commit I am happy with opening the pull request. I have made a small refactoring based on comments which I think is an improvement from what I had before. I think some of the stuff related to checking the header can moved to a utils file (not sure what to call it).

Regarding tests, I wouldn't minding doing them as there are still likely corner cases that I did not manually test or don't know about. However, as we might have merge conflicts and if you think we can make this release, I'm ok with you taking from here.

File: builtin/logical/pki/config_util.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (cipherboy) [reply builtin/logical/pki/config_util.go:60]: I might be missing something .... but `updateDefaultIssuerId` is called:

 1. During `config/issuer` operations, where its the last call (`path_config_ca.go`). So `issuer.LastModified` assignment without a `writeIssuer` call in this case will result in the modification not being persisted.
 2. When there's a new key (via `importKey`) and we lack a default issuer but have an issuer using this key. Then we'll call this, but again, there's no subsequent call to `writeIssuer`.
 3. Same holds for `importIssuer`.

However, these are kinda moot as the issue is locally scoped: the issuer is fetched in this function but not returned from this function, so the modifications will disappear at the end of the function return if they aren't persisted?

If you use `/sys/raw` [to read the stored issuer config](https://www.vaultproject.io/api-docs/system/raw) I think you'll see its not being persisted.

I'd imagine that the following would be enough to verify this:

1. Create two issuers, A and B.
2. Assign default to A.
3. Fetch with the header via `/cert/ca`. Use the current timestamp.
4. Wait a bit.
5. Set the default to B (`/config/issuers`).
6. Wait a bit.
7. Fetch with the header via `/cert/ca`. Use the timestamp of the last fetch, not right now. 

I think 7 will yield A, vs. B. Because while the issuer was modified since there, this write isn't issued. My 2c.?

[the remark this answers is not in the document]
Comment (Gabrielopesantos) [reply builtin/logical/pki/config_util.go:60]: You are right, if `writeIssuer` isn't called here, when we change the default ref the issuer LastModified field isn't updated.

[the remark this answers is not in the document]
Comment (Gabrielopesantos) [reply builtin/logical/pki/config_util.go:60]: Yes, you are right. If we don't writeIssuer the LastModified update isn't persisted.

[the remark this answers is not in the document]
Comment (cipherboy) [reply builtin/logical/pki/config_util.go:60]: Cool, ty!

File: builtin/logical/pki/util.go [hunk 12]
```diff
@@ -227,3 +229,95 @@ func isStringArrayDifferent(a, b []string) bool {
...
+func sendNotModifiedResponseIfNecessary(helper *IfModifiedSinceHelper, sc *storageContext, resp *logical.Response) (bool, error) {
+	var ret bool
+	responseHeaders := map[string][]string{}
+	if !hasHeader(headerIfModifiedSince, helper.req) {
+		return ret, nil
+	}
+
+	before, err := sc.isIfModifiedSinceBeforeLastModified(helper, responseHeaders)
+	if err != nil {
+		ret = true
```
Comment (stevendpclark) [inline builtin/logical/pki/util.go:269]: nit: Curious why are we setting this to true in the error case, I would expect this to be false as we haven't setup the headers like we do in the normal true case.

File: builtin/logical/pki/util.go [hunk 13]
```diff
@@ -227,3 +229,95 @@ func isStringArrayDifferent(a, b []string) bool {
...
+		return ret, err
+	}
+	if !before {
+		return ret, nil
+	} else {
```
Comment (stevendpclark) [inline builtin/logical/pki/util.go:274]: nit: No need for this else

File: builtin/logical/pki/util.go [hunk 14]
```diff
@@ -227,3 +229,95 @@ func isStringArrayDifferent(a, b []string) bool {
...
+	return headerTimeValue, nil
+}
+
+type IfModifiedSinceHelper struct {
+	req       *logical.Request
+	issuerRef issuerID
+}
+
+func sendNotModifiedResponseIfNecessary(helper *IfModifiedSinceHelper, sc *storageContext, resp *logical.Response) (bool, error) {
+	var ret bool
```
Comment (stevendpclark) [inline builtin/logical/pki/util.go:261]: nit: Technically we don't need this var at all, the boolean values could just be hardcoded to the values we want to return.

**Review by stevendpclark — APPROVED** [review]
A few nits but hardly worth keeping this from being merged.

**Review by stevendpclark — APPROVED** [review]
👍

**Comment by cipherboy** [comment]
A few follow-up minor changes can be merged in the next PR.

Thank you @Gabrielopesantos for kicking this off and doing a lot of the work in this PR!!
